### ISSUE TITLE
"no tests found" with AndroidJUnitRunner on Lollipop and Marshmallow

### ISSUE THREAD

**Issue by kkovach** [google/dagger#271]
I have some tests in an Android library that run just fine on Jellybean and Kitkat but fail when I try to run them on either Lollipop or Marshmallow. I've been unsuccessful in finding anyone else that's experiencing this behavior. It's pretty strange and confusing so I put together a little repo to try and understand and demonstrate what I'm experiencing here, https://github.com/kkovach/marshmallow-testing The README has all of the info I've collected thus far.

tldr; merely adding Dagger 2.0.2 as a dependency appears to cause AndroidJUnitRunner not to see tests and connectedCheck to fail.

I'd appreciate any info or advice anyone has if I'm missing something. Thanks.

- Kevin

**Comment by artem-zinnatullin** [comment]
The problem is that you [*compile*](https://github.com/kkovach/marshmallow-testing/blob/1d30e6f675fdbeb871666df5fa5d15252e9ca9a4/library/build.gradle#L35) `dagger-compiler` to the project. I don't want to investigate why it creates problems with instrumentation tests on Android 5+ but anyway, just do not *compile* it to the apk :smile: 

// probably test runner uses Dagger 2 internally and you break it via compiling own version, probably they even use Dagger 1, package names are same.

You can use APT Gradle Plugin https://bitbucket.org/hvisser/android-apt to run `dagger-compiler` during the compilation but do not *compile* it to the app.

Here is the [#qualitymatters app](https://github.com/artem-zinnatullin/qualitymatters) with instrumentation tests and Dagger 2, everything works without problems, so you can take a look and see how to bake something similar.

**Comment by kkovach** [comment]
Interesting. Thanks for the information. This does indeed allow me to run tests on Lollipop and Marshmallow.

Just for completeness, in actuality this was a result of Dagger 2 (`dagger-compile`) being used as a `compile` dependency in a POJO library that I've written that is being used as a `compile` dependency in my Android library. So, it appears that I have to exclude dagger from it.

I'll have to do some research on how to better use the `dagger-compile` dependency in the POJO library and keep it from being compiled in there.

**Comment by JakeWharton** [comment]
If Gradle, use https://github.com/tbroyer/gradle-apt-plugin. It Maven, use `<optional>true</optional>`.

**Comment by artem-zinnatullin** [comment]
@kkovach great to hear that it solved your problem! Can you please close the issue then?

**Comment by kkovach** [comment]
Thanks guys. These look to be the pieces I was missing.
